conversant/prompt_chatbot.py

```diff
@@ -113,7 +124,58 @@ def latest_prompt(self) -> str:
         """
         return self.prompt_history[-1]
 
-    def reply(self, query: str) -> Dict[str, str]:
+    def update_max_context_examples(self, prompt_size: int) -> int:
+        """Adjust max_context_examples until a possible prompt size.
+
+        if this is not possible, send an error message.
+
+        Args:
+            prompt_size (str): Number of tokens of the prompt
+
+        Returns:
+           int: updated max_context_examples
+        """
+
+        # Store original values
+        original_size = prompt_size
+        original_max_context_examples = self.chatbot_config["max_context_examples"]
+
+        # Reduce max_context_examples until the number of token of the prompt
+        # is less than maximum or reaches 1
+        start_point = max(
+            (len(self.chat_history) - self.chatbot_config["max_context_examples"]), 0
+        )
+        end_point = min(
+            self.chatbot_config["max_context_examples"] + start_point,
+            len(self.chat_history),
+        )
+
+        for i in range(start_point, end_point):
+            prompt_size = prompt_size - self.chat_history_sizes[i]
+            if prompt_size <= self.max_prompt_size:
+                max_context_examples = end_point - 1 - i
```

Comment:
I think a more Pythonic setup here might be something like the following:

```python
# assuming there is a local max_context_examples int passed as an arg
# and a trimmed_max_examples which is the new value we compute

trimmed_max_examples = max_context_examples
for size in self.chat_history_sizes[-max_context_examples:]:
    prompt_size -= size
    trimmed_max_examples -= 1
   
    if prompt_size <= self.max_prompt_size:
        # log warning
        return trimmed_max_examples

# raise error if condition never reached
```

I haven't tested it so there might be mistakes, but the general idea is that the list slicing should do a lot of the heavy-lifting with checking the bounds and making sure you get the "trailing slice" of context without going out of bounds.